clients/js/src/das.ts

```diff
@@ -0,0 +1,318 @@
+    ...(frozen !== undefined ? { frozen } : {}),
+  }))(dasPluginData);
+}
+
+export function dasPluginsToCorePlugins(dasPlugins: Record<string, any>) {
+  return Object.keys(dasPlugins).reduce((acc: PluginsList, dasPluginKey) => {
+    const dasPlugin = dasPlugins[dasPluginKey];
+    const { authority, data, offset } = dasPlugin;
+    const authorityAddress = authority?.address;
+
+    acc = {
+      ...acc,
+      [convertSnakeCase(dasPluginKey)]: {
+        authority: {
+          type: authority.type,
+          ...(authorityAddress ? { address: publicKey(authorityAddress) } : {}),
+        },
+        ...dasPluginDataToCorePluginData(data),
+        offset: BigInt(offset),
+      },
+    };
+
+    return acc;
+  }, {});
+}
+
+export function dasAssetToCoreAssetOrCollection(
+  dasAsset: DasApiAsset
+): AssetV1 | CollectionV1 {
+  // TODO: Define types in Umi DAS client.
+  const {
+    interface: assetInterface,
+    id,
+    ownership: { owner },
+    content: {
+      metadata: { name },
+      json_uri: uri,
+    },
+    compression: { seq },
+    grouping,
+    authorities,
+    plugins,
+    executable,
+    lamports: lamps,
+    rent_epoch: rentEpoch,
+    mpl_core_info: mplCoreInfo,
+  } = dasAsset as DasApiAsset & {
+    plugins: Record<string, any>;
+    executable?: boolean;
+    lamports?: number;
+    rent_epoch?: number;
+    mpl_core_info?: {
+      num_minted: number;
+      current_supply: number;
+      plugins_json_version: number;
+    };
+  };
+  const { num_minted: numMinted = 0, current_supply: currentSize = 0 } =
+    mplCoreInfo ?? {};
+  const commonFields = {
+    publicKey: id,
+    uri,
+    name,
+    ...getAccountHeader(executable, lamps, rentEpoch),
+    ...dasPluginsToCorePlugins(plugins),
+    // pluginHeader: // TODO: Reconstruct
+  };
+
+  const isCollection =
+    assetInterface === (MPL_CORE_COLLECTION as DasApiAssetInterface);
+  if (isCollection) {
+    return {
+      ...commonFields,
+      key: Key.CollectionV1,
+      updateAuthority: authorities[0].address,
+      numMinted,
+      currentSize,
+    };
+  }
+
+  return {
+    ...commonFields,
+    key: Key.AssetV1,
+    owner,
+    seq: seq ? some(BigInt(seq)) : none(),
+    ...getUpdateAuthority(grouping[0], authorities[0]),
+  };
+}
+
+export async function searchAssets(
+  context: Umi,
+  input: Omit<SearchAssetsRpcInput, 'interface' | 'burnt'> & {
+    interface: typeof MPL_CORE_ASSET | typeof MPL_CORE_COLLECTION;
+  }
+) {
+  const dasAssets = await context.rpc.searchAssets({
+    ...input,
+    interface: input.interface as DasApiAssetInterface,
+    burnt: false,
+  });
+
+  return dasAssets.items.map((dasAsset) =>
+    dasAssetToCoreAssetOrCollection(dasAsset)
+  );
+}
+
+export async function fetchAssetsByOwner(
+  context: Umi,
+  input: {
+    owner: PublicKey;
+  } & Pagination
+) {
+  const assets = await searchAssets(context, {
+    interface: MPL_CORE_ASSET,
+    owner: input.owner,
+  });
+
+  if (isAssets(assets)) return assets;
```

Comment:
You're right. I've missed optional chaining there.

As for why I use it. To make TS happy. It's a guard clause. Since `searchAssets` returns  `Promise<(AssetV1 | CollectionV1)[]>`. I wanted to narrow the return type of `fetchAssetsBy...` functions. And I didn't want to go with something like `as unknown as AssetV1`

If you have better suggestions, I would gladly implement them.

Reply:
1. Removed type guards. Went with function overloads
2. Added `searchCollections`
3. Made the `interface` param optional with a default value for `searchAssets`
https://github.com/metaplex-foundation/mpl-core/pull/30/commits/126c1768c7dc0eb22dc33b484c39e02112c0cc16#diff-f5937e280a48633613487d6f8b6b01877b83f068692e1d1812e1078374074e77